WordPress/gutenberg
# [RNMobile] Add Featured Banner to Image Block (iOS Only)

Paritial fix for: wordpress-mobile/gutenberg-mobile#1011

`gutenberg-mobile`: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3390
`WordPress-iOS`: https://github.com/wordpress-mobile/WordPress-iOS/pull/16332

The main PR for the branch this individual PR will be merged into, `add/featured-functionality-to-image-block-ios`, can be found here: https://github.com/WordPress/gutenberg/pull/31345

## Description

This PR introduces a **Featured** banner, which overlays a post's featured image when it's added to an image block. This will make it easier for users to identify which of the images in their posts, if any, are set as featured directly from the editor.

_Note: This PR introduces the **Featured** banner on iOS devices only, building on the work done in #30806 to add the banner to Android devices._

## How has this been tested?

### Test Case 1: General Flow

To start with, an image block should be updated to reflect whether its image is featured or not when set via the general Post Settings flow:

* Tap the three icons to the upper right of the post editor to access **Post Settings** and set a featured image.
* Tap the back button. If the new featured image hasn't already been added to the editor via an image block, add it.  
* Confirm that the **Featured** banner overlays the correct image block.
* Return to **Post Settings**. Tap the existing featured image and then **Remove**.
* Tap the back button to confirm that the **Featured** banner has been removed from the image block, as expected.

### Test Case 2: "Editor Loads" Flow

Next, we should verify that a **Featured** banner appears when a post with a featured image is first loaded. To test this, follow these steps:

* Verify that a featured image has been set on the post and tap the back button.
* Return to the post's editor by tapping on it from the post list screen.
* Verify that the post's featured image has a **Featured** banner overlaying it.

### Test Case 3: "Replace in Image Block" Flow

Finally, it's possible to replace an image directly within the image block's settings. We need to verify that the **Featured** banner updates as expected depending on the image that's chosen for an image block:

* Select any image block that is flagged with the **Featured** banner.
* Tap the edit icon to the upper right and then **Replace**. 
* Replace with any image that _isn't_ the post's featured image.
* Verify that the **Featured** banner is removed from the block when the new image uploads.
* Go through the steps to replace the block's image again, this time pick the post's featured image.
* Verify that the **Featured** banner appears again.

## Screenshots 

| Before  | After |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/116595561-59c18d00-a91b-11eb-8cef-66b4fa9ff096.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/116595474-42829f80-a91b-11eb-99a9-efae331b2f83.png" width="100%"> |

## Types of changes

This PR introduces a new feature (non-breaking change which adds functionality).

In https://github.com/WordPress/gutenberg/pull/30806, changes were introduced to display a `Featured` banner on Android devices. That PR can be referred to for further details on [the badge component](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/mobile/badge) that's used for the banner and the logic for storing `featuredImageId` in [the image component's edit.native.js file](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/image/edit.native.js#L612). To build on those changes and display the banner on iOS devices, the following main changes were introduced to the code: 

* `gutenbergFeaturedImageId` is [passed over the bridge](https://github.com/WordPress/gutenberg/pull/31347/files#diff-2b710719f1d5f87c056086112cc54b86275facfe3cfa93d502a594261cf0ad0fR67) to gather any initial featured image ID when a post is first loads.
* An [event subscription](https://github.com/WordPress/gutenberg/pull/31347/files#diff-bf4c994cef3006848617e4cd8a8ad4bfbc79ef6feda4d58152369ec92019d51eR66) is used to [track a change to a post's featured image](https://github.com/WordPress/gutenberg/pull/31347/files#diff-2b710719f1d5f87c056086112cc54b86275facfe3cfa93d502a594261cf0ad0fR123) from the iOS app.

## Checklist:
